### PR TITLE
#41/#43 corrected output for enhancements US01 and US03

### PR DESCRIPTION
**Pull request by piglet25**

### CODE REVIEW

File: src/yanfeishao/cs555/constant/ErrorInfo.java [hunk 1]
```diff
@@ -39,7 +39,7 @@
     /**
      * The constant US10_FAMILY.
      */
-    public static final String US10_FAMILY = "Marriage should be at least 14 years after birth of both spouses";
+    public static final String US10_FAMILY = "Marriage should be at least 14 years after birth of both spouses, family [%s] wife age [%s] and husband age [%s] at marriage";
     /**
```
Comment (JackieDreamy) [inline src/yanfeishao/cs555/constant/ErrorInfo.java]: Please add more information for one person, like wife [Personal Id %s Name: %s]

File: src/yanfeishao/cs555/constant/ErrorInfo.java [hunk 2]
```diff
@@ -51,7 +51,7 @@
...
      * The constant US21_FAMILY.
      */
-    public static final String US21_FAMILY = "Husband in family should be male and wife in family should be female";
+    public static final String US21_FAMILY = "Husband in family [%s] should be male [%s] and wife in family should be female [%s]";
     /**
```
Comment (JackieDreamy) [inline src/yanfeishao/cs555/constant/ErrorInfo.java]: Same above

File: src/yanfeishao/cs555/utils/AttributeUtils.java [hunk 3]
```diff
@@ -34,10 +34,23 @@ private boolean isLegalLastName(PersonEntity child, FamilyEntity familyEntity) {
         return splitName(child.getName())[1].equals(splitName(familyEntity.getFather().getName())[1]);
     }
 
+    private boolean isLegalMother(FamilyEntity familyEntity) {
+        return familyEntity.getMother().getSex().equalsIgnoreCase("F");
```
Comment (JackieDreamy) [inline src/yanfeishao/cs555/utils/AttributeUtils.java]: Please add F and M to KeywordsConstant.F and .M

File: src/yanfeishao/cs555/utils/AttributeUtils.java [hunk 4]
```diff
@@ -34,10 +34,23 @@ private boolean isLegalLastName(PersonEntity child, FamilyEntity familyEntity) {
...
+    }
+
+    private boolean isLegalFather(FamilyEntity familyEntity) {
+        return familyEntity.getFather().getSex().equalsIgnoreCase("M");
+    }
+
```
Comment (JackieDreamy) [inline src/yanfeishao/cs555/utils/AttributeUtils.java]: Same above

File: src/yanfeishao/cs555/utils/AttributeUtils.java [hunk 5]
```diff
@@ -55,4 +68,21 @@ public void parseUS16Error(Set<String> result, String prefix, FamilyEntity famil
...
+     */
+    public void parseUS21Error(Set<String> result, String prefix, FamilyEntity familyEntity) {
+        if (us21ParseCondition(familyEntity)) {
+            if (CommonUtils.isNotNull(familyEntity.getMother()) && familyEntity.getMother().getSex().equalsIgnoreCase("M")) {
+                result.add(String.format(FormatterRegex.ERROR_PERSON + ErrorInfo.US21_PERSON, prefix, familyEntity.getIdentifier(), "wife", "male", familyEntity.getFather().getIdentifier(), familyEntity.getFather().getName(), familyEntity.getFather().getSex(), familyEntity.getMother().getIdentifier(), familyEntity.getMother().getName(), familyEntity.getMother().getSex()));
+            } else if (CommonUtils.isNotNull(familyEntity.getMother()) && familyEntity.getFather().getSex().equalsIgnoreCase("F")) {
+                result.add(String.format(FormatterRegex.ERROR_PERSON + ErrorInfo.US21_PERSON, prefix, familyEntity.getIdentifier(), "husband", "female", familyEntity.getFather().getIdentifier(), familyEntity.getFather().getName(), familyEntity.getFather().getSex(), familyEntity.getMother().getIdentifier(), familyEntity.getMother().getName(), familyEntity.getMother().getSex()));
+            }
```
Comment (JackieDreamy) [inline src/yanfeishao/cs555/utils/AttributeUtils.java]: Same above add to keywordConstant

File: src/yanfeishao/cs555/utils/DateUtils.java [hunk 6]
```diff
@@ -66,14 +72,18 @@ private void parseChildDateError(Set<String> result, FamilyEntity familyEntity,
...
                 switch (prefix) {
                     case ErrorCode.US01: {
+                        Date currentDate = CommonUtils.getCurrentDate();
                         if (lifeDateWithRange(childBirthDate, childDeathDate)) {
-                            result.add(String.format(FormatterRegex.ERROR_FAMILY + ErrorInfo.US01_FAMILY, prefix, familyEntity.getIdentifier()));
+                            if (CommonUtils.isNotNull(childBirthDate) && childBirthDate.after(currentDate))
+                                result.add(String.format(FormatterRegex.ERROR_PERSON + ErrorInfo.US01_PERSON, prefix, familyEntity.getIdentifier(), child.getIdentifier(), child.getName(), "Birth", CommonUtils.getFormattedDate(childBirthDate), CommonUtils.getFormattedDate(currentDate)));
+                            else if (CommonUtils.isNotNull(childDeathDate) && childDeathDate.after(CommonUtils.getCurrentDate()))
+                                result.add(String.format(FormatterRegex.ERROR_PERSON + ErrorInfo.US01_PERSON, prefix, familyEntity.getIdentifier(), child.getIdentifier(), child.getName(), "Death", CommonUtils.getFormattedDate(childDeathDate), CommonUtils.getFormattedDate(currentDate)));
```
Comment (JackieDreamy) [inline src/yanfeishao/cs555/utils/DateUtils.java]: Add Birth and Death to KeywordsConstant

File: src/yanfeishao/cs555/utils/DateUtils.java [hunk 7]
```diff
@@ -115,7 +125,20 @@ private void parseChildDateError(Set<String> result, FamilyEntity familyEntity,
...
+            } else if (CommonUtils.isNotNull(marriageDate) && marriageDate.after(currentDate)) {
+                result.add(String.format(FormatterRegex.ERROR_PERSON + ErrorInfo.US01_PERSON, prefix, familyEntity.getIdentifier(), familyEntity.getFather().getIdentifier() + " & " + familyEntity.getMother().getIdentifier(), familyEntity.getFather().getName() + " & " + familyEntity.getMother().getName(), "Marriage", CommonUtils.getFormattedDate(marriageDate), CommonUtils.getFormattedDate(currentDate)));
+            } else if (CommonUtils.isNotNull(divorceDate) && divorceDate.after(currentDate)) {
+                result.add(String.format(FormatterRegex.ERROR_PERSON + ErrorInfo.US01_PERSON, prefix, familyEntity.getIdentifier(), familyEntity.getFather().getIdentifier() + " & " + familyEntity.getMother().getIdentifier(), familyEntity.getFather().getName() + " & " + familyEntity.getMother().getName(), "Divorce", CommonUtils.getFormattedDate(divorceDate), CommonUtils.getFormattedDate(currentDate)));
```
Comment (JackieDreamy) [inline src/yanfeishao/cs555/utils/DateUtils.java]: Same above

File: src/yanfeishao/cs555/constant/ErrorInfo.java (hunk 1 above)
Comment (piglet25) [inline src/yanfeishao/cs555/constant/ErrorInfo.java]: This consant is no longer used it has been replaced with US10_Person 
that has the requested additional information.

File: src/yanfeishao/cs555/constant/ErrorInfo.java (hunk 2 above)
Comment (piglet25) [inline src/yanfeishao/cs555/constant/ErrorInfo.java]: constant US21_FAMILY is no longer used it is replaced with the constant 
US21_PERSON which has the requested information.

**Comment by JackieDreamy** [comment]
Piglet, please sync the branch before you send pull request. I will close this pull request, please fix above issue and send again

File: src/yanfeishao/cs555/constant/ErrorInfo.java (hunk 1 above)
Comment (JackieDreamy) [inline src/yanfeishao/cs555/constant/ErrorInfo.java]: Ok, that's fine

**Comment by piglet25** [comment]
I don't understand meaning of this. Please fix what? I fixed all 
constants and made new commit and pull request. If I do a sync won't I 
overwrite all my changes?

**Pull request closed without merging**